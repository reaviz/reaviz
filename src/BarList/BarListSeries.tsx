import React, { FC, useCallback } from 'react';
import {
  ChartInternalDataTypes,
  ChartShallowDataShape,
  ColorSchemeType,
  DEFAULT_TRANSITION,
  formatValue,
  getAriaLabel,
  getColor
} from '@/common';
import { motion } from 'motion/react';
import classNames from 'classnames';
import css from './BarListSeries.module.css';

export type BarListLabelPosition = 'none' | 'top' | 'start' | 'end' | 'bottom';

export interface BarListSeriesProps {
  /**
   * Data for the chart.
   */
  data?: ChartShallowDataShape[];

  /**
   * Color scheme for the chart.
   */
  colorScheme?: ColorSchemeType;

  /**
   * The label position.
   */
  labelPosition?: BarListLabelPosition;

  /**
   * The value position.
   */
  valuePosition?: BarListLabelPosition;

  /**
   * The bar item class name.
   */
  itemClassName?: string;

  /**
   * Label css class name.
   */
  labelClassName?: string;

  /**
   * Label value class name.
   */
  valueClassName?: string;

  /**
   * Bar component class name.
   */
  barClassName?: string;

  /**
   * Bar container class name.
   */
  outerBarClassName?: string;

  /**
   * Custom label format.
   */
  labelFormat?: (data: ChartInternalDataTypes, index: number) => any;

  /**
   * Custom value format
   */
  valueFormat?: (data: ChartInternalDataTypes, index: number) => any;

  /**
   * Item was clicked.
   */
  onItemClick?: (data: ChartShallowDataShape) => void;

  /**
   * Item had mouse enter.
   */
  onItemMouseEnter?: (data: ChartShallowDataShape) => void;

  /**
   * Item had mouse leave.
   */
  onItemMouseLeave?: (data: ChartShallowDataShape) => void;
}

export const BarListSeries: FC<Partial<BarListSeriesProps>> = ({
  data,
  colorScheme = 'cybertron',
  itemClassName,
  labelClassName,
  outerBarClassName,
  valueClassName,
  labelFormat,
  barClassName,
  labelPosition = 'top',
  valuePosition = 'none',
  valueFormat,
  onItemClick,
  onItemMouseEnter,
  onItemMouseLeave
}) => {
  const renderBar = useCallback(
    (item, index) => {
      const fill = getColor({
        data,
        colorScheme,
        point: item.data,
        index
      });

      return (
        <div className={classNames(css.outerBar, outerBarClassName)}>
          <motion.div
            transition={DEFAULT_TRANSITION}
            className={classNames(css.bar, barClassName)}
            initial={{ width: '0%' }}
            animate={{ width: `${item.data}%` }}
            style={{ background: fill }}
            tabIndex={0}
            aria-label={getAriaLabel(item)}
            role="graphics-document"
          />
        </div>
      );
    },
    [barClassName, outerBarClassName, colorScheme, data]
  );

  return (
    <>
      {data.map((d, i) => {
        const label = labelFormat ? labelFormat(d.key as any, i) : d.key;

        const valueLabel = valueFormat
          ? valueFormat(d.metadata.value, i)
          : formatValue(d.metadata.value);

        return (
          <div
            key={d.key as string}
            role="listitem"
            className={classNames(css.item, itemClassName, {
              [css.clickable]: onItemClick,
              [css.valueBottom]: valuePosition === 'bottom',
              [css.valueStart]: valuePosition === 'start',
              [css.valueEnd]: valuePosition === 'end',
              [css.valueNone]: valuePosition === 'none',
              [css.labelBottom]: labelPosition === 'bottom',
              [css.labelTop]: labelPosition === 'top',
              [css.labelStart]: labelPosition === 'start',
              [css.labelEnd]: labelPosition === 'end',
              [css.labelNone]: labelPosition === 'none'
            })}
            onMouseEnter={() => onItemMouseEnter?.(d)}
            onMouseLeave={() => onItemMouseLeave?.(d)}
            onClick={() => onItemClick?.(d)}
          >
            <label
              title={label}
              className={classNames(css.label, labelClassName)}
            >
              {label}
            </label>
            {renderBar(d, i)}
            <label
              title={valueLabel}
              className={classNames(css.valueLabel, valueClassName)}
            >
              <small>{valueLabel}</small>
            </label>
          </div>
        );
      })}
    </>
  );
};
