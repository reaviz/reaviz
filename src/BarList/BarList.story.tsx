import { BarList } from './BarList';
import { BarListSeries } from './BarListSeries';

export default {
  tags: ['snapshot'],
  title: 'Charts/Bar List',
  component: BarList
};

export const Simple = () => (
  <BarList
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
  />
);

export const Styling = () => (
  <>
    <style>
      {`
        .bar {
          border-radius: 5px;
        }
        .outer {
          background: rgba(0, 0, 0, .2);
          border-radius: 5px;
        }
        .value {
          opacity: .5;
        }
      `}
    </style>
    <BarList
      id="styling"
      style={{ width: 350 }}
      data={[
        { key: 'Vulnerability Patch', data: 50 },
        { key: 'Critical Failure', data: 25 },
        { key: 'Physical Intrusion', data: 5 },
        { key: 'Phishing Attempts', data: 100 }
      ]}
      series={
        <BarListSeries
          barClassName="bar"
          outerBarClassName="outer"
          valueClassName="value"
          valuePosition="end"
        />
      }
    />
  </>
);

export const Events = () => (
  <>
    <BarList
      id="events"
      style={{ width: 350 }}
      data={[
        { key: 'Vulnerability Patch', data: 50 },
        { key: 'Critical Failure', data: 25 },
        { key: 'Physical Intrusion', data: 5 },
        { key: 'Phishing Attempts', data: 100 }
      ]}
      series={
        <BarListSeries
          onItemClick={(d) => alert(`Clicked ${d.key}`)}
          outerBarClassName="outer"
          valuePosition="end"
        />
      }
    />
  </>
);

export const PercentageFormat = () => (
  <>
    <style>
      {`
        .bar {
          border-radius: 5px;
        }
        .outer {
          background: rgba(0, 0, 0, .2);
          border-radius: 5px;
        }
        .value {
          width: 40px;
        }
      `}
    </style>
    <BarList
      id="percentage-format"
      style={{ width: 350 }}
      data={[
        { key: 'Vulnerability Patch', data: 50 },
        { key: 'Critical Failure', data: 25 },
        { key: 'Physical Intrusion', data: 5 },
        { key: 'Phishing Attempts', data: 85 }
      ]}
      series={
        <BarListSeries
          valueFormat={(data) => `${data}%`}
          onItemClick={(d) => alert(`Clicked ${d.key}`)}
          valuePosition="end"
          valueClassName="value"
          barClassName="bar"
          outerBarClassName="outer"
        />
      }
      type="percent"
    />
  </>
);

export const Empty = () => (
  <>
    <style>
      {`
        .bar {
          border-radius: 5px;
        }
        .outer {
          background: rgba(0, 0, 0, .2);
          border-radius: 5px;
        }
        .value {
          opacity: .5;
        }
      `}
    </style>
    <BarList
      id="empty"
      style={{ width: 350 }}
      data={[
        { key: 'Vulnerability Patch', data: 0 },
        { key: 'Critical Failure', data: 0 },
        { key: 'Physical Intrusion', data: 0 },
        { key: 'Phishing Attempts', data: 0 }
      ]}
      series={
        <BarListSeries
          barClassName="bar"
          outerBarClassName="outer"
          valueClassName="value"
          valuePosition="end"
        />
      }
    />
  </>
);
