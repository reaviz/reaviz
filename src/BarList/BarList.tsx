import { max } from 'd3-array';
import { scaleLinear } from 'd3-scale';
import { motion } from 'motion/react';
import { CloneElement, useId } from 'reablocks';
import type { FC, ReactElement } from 'react';
import React, { useMemo } from 'react';

import type { ChartShallowDataShape } from '@/common';

import type { BarListSeriesProps } from './BarListSeries';
import { BarListSeries } from './BarListSeries';

export interface BarListProps {
  /**
   * ID for the chart.
   */
  id?: string;

  /**
   * CSS Classname for the element.
   */
  className?: string;

  /**
   * Style for the element.
   */
  style?: React.CSSProperties;

  /**
   * Data for the chart.
   */
  data: ChartShallowDataShape[];

  /**
   * Sort direction of the data.
   */
  sortDirection?: 'asc' | 'desc' | 'none';

  /**
   * Series to render.
   */
  series?: ReactElement<BarListSeriesProps, typeof BarList>;

  /**
   * Whether the values are percentages or absolute values.
   * In the latter case, the chart would be relative
   */
  type?: 'percent' | 'count';
}

export const BarList: FC<BarListProps> = ({
  id,
  className,
  style,
  data = [],
  sortDirection = 'desc',
  series = <BarListSeries />,
  type = 'count',
}) => {
  const curId = useId(id);

  const mashedData = useMemo(() => {
    const maxVal = type === 'count' ? max(data, (d) => d.data as number) : 100;
    const domainVal = maxVal === 0 ? [0] : [0, maxVal];
    const groupScale = scaleLinear()
      .domain(domainVal as [number, number])
      .rangeRound([0, 100]);

    const mashed = data.map((d) => ({
      ...d,
      data: groupScale(d.data as number),
      metadata: {
        value: d.data,
        percent: (data.length * (d.data as number)) / 100,
      },
    }));

    if (sortDirection === 'asc') {
      mashed.sort((a, b) => a.data - b.data);
    } else if (sortDirection === 'desc') {
      mashed.sort((a, b) => b.data - a.data);
    }

    return mashed;
  }, [data, sortDirection]);

  return (
    <motion.section
      role="list"
      variants={{
        initial: {
          transition: {
            staggerChildren: 0.05,
            staggerDirection: -1 as any,
          },
        },
        animate: {
          transition: {
            staggerChildren: 0.07,
            delayChildren: 0.2,
          },
        },
      }}
      id={curId}
      className={className}
      style={style}
    >
      <CloneElement<BarListSeriesProps> element={series} data={mashedData} />
    </motion.section>
  );
};
