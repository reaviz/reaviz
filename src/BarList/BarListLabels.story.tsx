import React from 'react';
import { BarList } from './BarList';
import { BarListSeries } from './BarListSeries';

export default {
  tags: ['snapshot'],
  title: 'Charts/Bar List/Labels',
  component: BarList
};

export const LabelStart = () => (
  <BarList
    id="label-start"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
    series={<BarListSeries labelPosition="start" valuePosition="end" />}
  />
);

export const LabelBottom = () => (
  <BarList
    id="label-bottom"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
    series={<BarListSeries labelPosition="bottom" valuePosition="end" />}
  />
);

export const LabelEnd = () => (
  <BarList
    id="label-end"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
    series={<BarListSeries labelPosition="end" valuePosition="start" />}
  />
);

export const LabelTop = () => (
  <BarList
    id="label-top"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
    series={<BarListSeries labelPosition="top" />}
  />
);

export const LabelNone = () => (
  <BarList
    id="label-none"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
    series={<BarListSeries labelPosition="none" valuePosition="end" />}
  />
);

export const ValueBottom = () => (
  <BarList
    id="value-bottom"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
    series={<BarListSeries valuePosition="bottom" />}
  />
);

export const ValueEnd = () => (
  <BarList
    id="value-end"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 100 },
      { key: 'Critical Failure', data: 55 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 200 }
    ]}
    series={<BarListSeries valuePosition="end" />}
  />
);

export const ValueStart = () => (
  <BarList
    id="value-start"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
    series={<BarListSeries valuePosition="start" />}
  />
);

export const ValueNone = () => (
  <BarList
    id="value-none"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
    series={<BarListSeries valuePosition="none" />}
  />
);

export const LongValueStart = () => (
  <BarList
    id="long-value-start"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 5000 },
      { key: 'Critical Failure', data: 50000 },
      { key: 'Physical Intrusion', data: 500000 },
      { key: 'Phishing Attempts', data: 5000000 }
    ]}
    series={<BarListSeries valuePosition="start" />}
  />
);

export const LongValueEnd = () => (
  <BarList
    id="long-value-end"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 5000 },
      { key: 'Critical Failure', data: 50000 },
      { key: 'Physical Intrusion', data: 500000 },
      { key: 'Phishing Attempts', data: 5000000 }
    ]}
    series={<BarListSeries valuePosition="end" />}
  />
);
