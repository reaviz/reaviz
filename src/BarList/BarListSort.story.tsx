import React from 'react';
import { BarList } from './BarList';

export default {
  tags: ['snapshot'],
  title: 'Charts/Bar List/Sort',
  component: BarList
};

export const Ascending = () => (
  <BarList
    id="ascending"
    sortDirection="asc"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
  />
);

export const Descending = () => (
  <BarList
    id="descending"
    sortDirection="desc"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
  />
);

export const None = () => (
  <BarList
    id="none"
    sortDirection="none"
    style={{ width: 350 }}
    data={[
      { key: 'Vulnerability Patch', data: 50 },
      { key: 'Critical Failure', data: 25 },
      { key: 'Physical Intrusion', data: 5 },
      { key: 'Phishing Attempts', data: 100 }
    ]}
  />
);
