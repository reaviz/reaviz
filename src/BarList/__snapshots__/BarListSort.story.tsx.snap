// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts/Bar List/Sort Ascending smoke-test 1`] = `
<section role="list"
         id="ascending"
         style="width: 350px;"
>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Physical Intrusion"
           class="_label_18e6r_8"
    >
      Physical Intrusion
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Physical Intrusion: 5"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 5%;"
      >
      </div>
    </div>
    <label title="5"
           class="_valueLabel_18e6r_71"
    >
      <small>
        5
      </small>
    </label>
  </div>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Critical Failure"
           class="_label_18e6r_8"
    >
      Critical Failure
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Critical Failure: 25"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 25%;"
      >
      </div>
    </div>
    <label title="25"
           class="_valueLabel_18e6r_71"
    >
      <small>
        25
      </small>
    </label>
  </div>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Vulnerability Patch"
           class="_label_18e6r_8"
    >
      Vulnerability Patch
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Vulnerability Patch: 50"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 50%;"
      >
      </div>
    </div>
    <label title="50"
           class="_valueLabel_18e6r_71"
    >
      <small>
        50
      </small>
    </label>
  </div>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Phishing Attempts"
           class="_label_18e6r_8"
    >
      Phishing Attempts
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Phishing Attempts: 100"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 100%;"
      >
      </div>
    </div>
    <label title="100"
           class="_valueLabel_18e6r_71"
    >
      <small>
        100
      </small>
    </label>
  </div>
</section>
`;

exports[`Charts/Bar List/Sort Descending smoke-test 1`] = `
<section role="list"
         id="descending"
         style="width: 350px;"
>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Phishing Attempts"
           class="_label_18e6r_8"
    >
      Phishing Attempts
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Phishing Attempts: 100"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 100%;"
      >
      </div>
    </div>
    <label title="100"
           class="_valueLabel_18e6r_71"
    >
      <small>
        100
      </small>
    </label>
  </div>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Vulnerability Patch"
           class="_label_18e6r_8"
    >
      Vulnerability Patch
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Vulnerability Patch: 50"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 50%;"
      >
      </div>
    </div>
    <label title="50"
           class="_valueLabel_18e6r_71"
    >
      <small>
        50
      </small>
    </label>
  </div>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Critical Failure"
           class="_label_18e6r_8"
    >
      Critical Failure
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Critical Failure: 25"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 25%;"
      >
      </div>
    </div>
    <label title="25"
           class="_valueLabel_18e6r_71"
    >
      <small>
        25
      </small>
    </label>
  </div>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Physical Intrusion"
           class="_label_18e6r_8"
    >
      Physical Intrusion
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Physical Intrusion: 5"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 5%;"
      >
      </div>
    </div>
    <label title="5"
           class="_valueLabel_18e6r_71"
    >
      <small>
        5
      </small>
    </label>
  </div>
</section>
`;

exports[`Charts/Bar List/Sort None smoke-test 1`] = `
<section role="list"
         id="none"
         style="width: 350px;"
>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Vulnerability Patch"
           class="_label_18e6r_8"
    >
      Vulnerability Patch
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Vulnerability Patch: 50"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 50%;"
      >
      </div>
    </div>
    <label title="50"
           class="_valueLabel_18e6r_71"
    >
      <small>
        50
      </small>
    </label>
  </div>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Critical Failure"
           class="_label_18e6r_8"
    >
      Critical Failure
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Critical Failure: 25"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 25%;"
      >
      </div>
    </div>
    <label title="25"
           class="_valueLabel_18e6r_71"
    >
      <small>
        25
      </small>
    </label>
  </div>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Physical Intrusion"
           class="_label_18e6r_8"
    >
      Physical Intrusion
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Physical Intrusion: 5"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 5%;"
      >
      </div>
    </div>
    <label title="5"
           class="_valueLabel_18e6r_71"
    >
      <small>
        5
      </small>
    </label>
  </div>
  <div role="listitem"
       class="_item_18e6r_1 _valueNone_18e6r_65 _labelTop_18e6r_42"
  >
    <label title="Phishing Attempts"
           class="_label_18e6r_8"
    >
      Phishing Attempts
    </label>
    <div class="_outerBar_18e6r_97">
      <div class="_bar_18e6r_105"
           tabindex="0"
           aria-label="Phishing Attempts: 100"
           role="graphics-document"
           style="background: rgb(41, 181, 202); width: 100%;"
      >
      </div>
    </div>
    <label title="100"
           class="_valueLabel_18e6r_71"
    >
      <small>
        100
      </small>
    </label>
  </div>
</section>
`;
